Add input plugin for OpenLDAP

cobaugh opened the pull request:
This is a plugin to pull metrics out of cn=Monitor from a local or remote OpenLDAP server.
### Required for all PRs:

- [ ] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [x] README.md updated (if adding a new plugin)

danielnelson commented:
Can you add some unit tests?

cobaugh commented:
Added unit tests. There is a very simple test function that mocks up a SearchRequest and feeds that in, and the integration tests connect to a super simple openldap container.

phemmer commented on plugins/inputs/openldap/README.md:
````diff
@@ -0,0 +1,76 @@
+```
+$ telegraf -config telegraf.conf -input-filter openldap -test --debug
+* Plugin: inputs.openldap, Collection 1
+> openldap,port=389,host=localhost,server=localhost abandon_operations_initiated=4,extended_operations_completed=125963,bytes_statistics=595939321,pdu_statistics=17028251,modify_operations_initiated=0,delete_operations_completed=0,compare_operations_completed=0,max_file_descriptors_connections=4096,unbind_operations_completed=7981688,extended_operations_initiated=125963,referrals_statistics=0,mod ...
````
Why are all the values being added as floats? None of them look to be floatish (have a decimal in them).

cobaugh replied:
You're right, the syntax for the monitorCounter attribute specifies a 32-bit integer. I'll go ahead and make that change.

phemmer commented on plugins/inputs/openldap/openldap.go:
```diff
@@ -0,0 +1,148 @@
+	TlsSkipverify bool
+	BindDn        string
+	BindPassword  string
+}
```
We should probably support SSL as well (meaning LDAP over SSL on port 636).

phemmer replied:
Also, what about providing a path to a custom CA cert, so that users running their own cert aren't forced to use tls_skipverify?

cobaugh replied:
I didn't implement ldapS because it was never formally standardized, LDAPv3 doesn't specify anything regarding ldaps, only starttls, and go-ldap itself doesn't directly support ldaps.

I'll have to dig into the tls and x509 packages to see about adding a ca cert option.

phemmer commented on plugins/inputs/openldap/openldap.go:
```diff
@@ -0,0 +1,148 @@
+	return nil
+}
+
+func gatherSearchResult(sr *ldap.SearchResult, o *Openldap, acc telegraf.Accumulator) error {
```
This function has an `error` in the return signature, but it doesn't ever return an error, nor is the caller checking the error.

phemmer commented on plugins/inputs/openldap/openldap.go:
```diff
@@ -0,0 +1,148 @@
+	return nil
+}
+
+// Convert a DN to metric name, eg cn=Read,cn=Waiters,cn=Monitor to read_waiters
```
What are your thoughts on reversing this? So that it's `waiters_read`? LDAP DNs are backwards, in that they go from most specific to least specific. The advantage of reversing this is that it allows grouping when sorted.
Meaning if you have `cn=Write,cn=Waiters,...`, `cn=Read,cn=Waiters,...`, `cn=Write,cn=foo,...`, and `cn=Read,cn=foo,...`, you end up with `foo_read,foo_write,waiters_read,waiters_write` instead of `read_foo,read_waiters,write_foo,write_waiters`.

cobaugh replied:
I debated having a map of DNs to field names instead. It does bother my OCD just a bit having read_waiters instead of waiters_read, but it also reads a little easier? If other plugins name their fields so they sort more naturally I can change this one to match that pattern.

cobaugh commented:
Added a tls_cacertificate option - not sure how best to do the integration test for that.
Field values are stored as ints now.
gatherSearchResults() returns nil.

I'm still a bit hesitant to support ldapS, but I will if there is consensus around it.

danielnelson commented:
@cobaugh Do you think the config can be made to look more like some the existing tls options?  Here is an example: https://github.com/influxdata/telegraf/blob/06baf7cf78bbdc2cd04ec141f2603498567b131a/plugins/inputs/mongodb/mongodb.go#L24

cobaugh commented:
@danielnelson In this case, since we aren't doing ssl, isn't tls_* more appropriate? Or do you mean just in general? eg:
tls_ca
insecure_skip_verify

>

danielnelson commented:
Probably should just call them `ssl_` since all the rest are misnamed.  If it all looks similar it will be easier to do a refactor later to rename them.

cobaugh commented:
I think it would be confusing to name them ssl_ when we we're doing ssl.

danielnelson commented:
I think being consistent is more important, most people still refer to it incorrectly as SSL anyway and I doubt anyone who knows how to use it correctly will be confused.  The more important part though is hopefully leveraging `internal.GetTLSConfig()`.

If you are interested we could work together on a concrete plan to migrate to new variable names across Telegraf, it is likely a lot of repetitious work but still important.

cobaugh commented:
Gotcha. I'll work on renaming the options and switching to internal.GetTLSConfig().

I'd also be happy to help with tls/ssl variable names where I can.

cobaugh commented:
I've converted the option names to be more in line with other plugins, and switched to using internal.GetTLSConfig(). My only hangup is with the ssl boolean itself. It seems wrong to name this option "ssl" to enable StartTLS. Perhaps I need to have another boolean called "starttls" and just make them mutually exclusive? I don't see any other plugins that seem to support both ssl and starttls, so I'm not entirely sure what to do here. Thoughts?

danielnelson commented:
The way is generally works in other plugins is that when the "ssl" options are unset, GetTLSConfig returns nil and when that gets passed into the Dial function it disables TLS.  So we don't usually need a flag for enabling it as it's implied by the values of the ssl options.

cobaugh commented:
Sorry for the delay. 

I'm not using the DialTLS function from ldap, but instead using StartTLS. I don't think DialTLS actually sends the STARTTLS command. I also don't think StartTLS will "just work" if you pass in a tlsConfig like DialTLS does. So, what semantics do we want for this? If we do want to support LDAPS in the future, we'll need some way of differentiating between starttls and ssl besides the ssl_ca config option being set.

danielnelson commented:
The STARTTLS vs ldaps bit did not really sink in for me before, but I think I got it now.

> I also don't think StartTLS will "just work" if you pass in a tlsConfig like DialTLS does.

Isn't the code currently passing a tls.Config into StartTLS?  Does it work now?

> If we do want to support LDAPS in the future, we'll need some way of differentiating between starttls and ssl besides the ssl_ca config option being set.

What if we use a origin style connection url?  By this I'm referring to a url of the form `scheme://host:port`.  This way you can control ldap vs ldaps via the scheme.  With a ldap url with certs it would use the starttls method.

cobaugh commented:
> Isn't the code currently passing a tls.Config into StartTLS? Does it work now?

Yes, and it works. What I meant was that if you call StartTLS, then you do StartTLS, period. So we need some way to know if the user wants us to do starttls or ssl.

> What if we use a origin style connection url? By this I'm referring to a url of the form scheme://host:port. This way you can control ldap vs ldaps via the scheme. With a ldap url with certs it would use the starttls method.

There was some discussion on the openldap mailing list over a decade ago about supporting a uri scheme for specifying starttls. ldaps://host:port obviously works for ldaps, but there does not currently exist a way to specify starttls in the uri.

I'm propsing three options here:
1) We have two boolean options, "ssl" and "starttls", mutually exclusive. Setting both to true would be a configuration error.
2) We do what some other projects do. We have one boolean option, "ssl". If ssl is true and port is 389, then we do starttls. If ssl is true and port is 636, then we do ssl. 
3) We have a single option that accepts either "ssl" or "starttls" as a string. Not a fan of this option.

Drawback to (2) is we can't easily support ldap+starttls or ldaps over any other port than the standard ports used with those protocols, though it does make the configuration real simple.

phemmer commented:
Well I seem to have opened a can of worms...
Since LDAP over SSL is not the preferred method of encryption, I'm perfectly fine with saying "won't do", or at least "won't do for now".
However that said, I think implementation would be best if it followed the openldap CLI utils method of differentiating SSL vs StartTLS. Meaning accept a URI of the form `ldap://...` and `ldaps://...`. When using `ldap://` have a separate option for enabling StartTLS. This option could just be a `tls_reqcert` option (which would behave like openldap's `TLS_REQCERT`). The presence of the config param, no matter what value, would enable StartTLS. And when used with `ldaps://...`, it would set the enforcement level.

This would make the configuration familiar to those already familiar with OpenLDAP, while keeping the number of config params to a minimum.

danielnelson commented:
I agree, URI plus a single boolean seems preferable to multiple booleans.

phemmer commented:
Well in OpenLDAP, `TLS_REQCERT` isn't a boolean. It accepts multiple values: 

> This option is set to never by default, in which case the server never asks the client for a certificate. With a setting of allow the server will ask for a client certificate; if none is provided the session proceeds normally. If a certificate is provided but the server is unable to verify it, the certificate is ignored and the session proceeds normally, as if no certificate had been provided. With a setting of try the certificate is requested, and if none is provided, the session proceeds normally. If a certificate is provided and it cannot be verified, the session is immediately terminated. With a setting of demand the certificate is requested and a valid certificate must be provided, otherwise the session is immediately terminated.

danielnelson commented:
Is this a server setting?

phemmer commented:
Client and server. The documentation I copied was the server side. The client side documentation just says "see server side documentation".

danielnelson commented:
Sounds good, lets just do it like OpenLDAP.

phemmer commented:
Do we want to do LDAPS on the first release, or just make sure that we would be able to add support in the future without breaking compatibility?

danielnelson commented:
So long as we can add it in the future I think we are good.

cobaugh commented:
Again, sorry to drag this out, but I'm still confused what the consensus is on this. TLS_REQCERT has to do with verifying the server certificate in the client context, and nothing to do with whether we are doing LDAPS or LDAP+StartTLS. Normally LDAPS is requested with a URI of ldaps://, and StartTLS is requested with the -ZZ option. To me, that means having a boolean called 'starttls'. 

If we want to take this conversation off of this issue to avoid cluttering it up I'm fine with that too.

Fwiw, grafana has two booleans for its ldap authentication, use_ssl and start_tls:
https://github.com/grafana/grafana/issues/2441

use_ssl enables ssl, and if start_tls is True, then it does a StartTLS instead of DialTLS.

danielnelson commented:
What happens if the `TLS_REQCERT` option is set to `demand`, but the `-ZZ` option is not given?  I think we will want to behave as closely as possible to the openldap client.

One common issue with using multiple booleans like grafana is that you could specify an invalid state combination, such as setting both start_tls and use_ssl to be true, or they could conflict with the url scheme which I think we will want to match openldap.

Don't worry having too much discussion here, this is the right place.  Best to sort this out beforehand than get stuck with a config file we don't like.

cobaugh commented:
If TLS_REQCERT is set to demand (the default), and neither -ZZ nor a URI of
ldaps:// is specified, it has no effect. The only way you can request
StartTLS with the cli tools with -Z(Z), and the only way you can request
ldaps is with the ldaps uri. If you specify -ZZ and ldaps://, you get an
error along the lines of "TLS already started".

With grafana setting both se_ssl = true and start_tls = true is how you
specify that you want to do StartTLS. There is no combination of those two
booleans that doesn't make sense. Grafana also doesn't allow one to specify
a connection url, only a hostname or IP. go-ldap also doesn't accept URLs,
so that's something we would need to handle ourselves.

On Mon, Jun 26, 2017 at 2:50 PM, Daniel Nelson <notifications@github.com>
wrote:

> What happens if the TLS_REQCERT option is set to demand, but the -ZZ
> option is not given? I think we will want to behave as closely as possible
> to the openldap client.
>
> One common issue with using multiple booleans like grafana is that you
> could specify an invalid state combination, such as setting both start_tls
> and use_ssl to be true, or they could conflict with the url scheme which I
> think we will want to match openldap.
>
> Don't worry having too much discussion here, this is the right place. Best
> to sort this out beforehand than get stuck with a config file we don't like.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/influxdata/telegraf/pull/2612#issuecomment-311148577>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAG57OaNHscnibq4Kbr_fkNqjWKStyBRks5sH_1ogaJpZM4Mxof9>
> .
>



-- 
andy

danielnelson commented:
The problem is that the options are mutually exclusive, since you would never do a tls connection and then issue a start_tls command.  So Grafana must be giving more precedence to one of them, I assume `use_ssl`.

One issue with using `ldap://` urls, if we do this I'm sure people will want to use all features of this url, and not understand that it is only meant to be an host + port.  `ldap://host:port/DN?attributes?scope?filter?extensions`.

Also, what about unix sockets, are these commonly used?

How do you feel about a single option that holds either ssl/tls or start_tls?

cobaugh commented:
Grafana's options are not mutually exclusive:
https://github.com/grafana/grafana/blob/master/pkg/login/ldap.go#L65

I also don't think we should support URLs unless we're willing to parse any
valid ldap url, like you suggested.

Sockets (ldapi://) are used sometimes in the wild. go-ldap doesn't seem to
support those either.

I'm fine with a single option that takes a string."ssl" or "starttls".

On Mon, Jun 26, 2017 at 5:27 PM, Daniel Nelson <notifications@github.com>
wrote:

> The problem is that the options are mutually exclusive, since you would
> never do a tls connection and then issue a start_tls command. So Grafana
> must be giving more precedence to one of them, I assume use_ssl.
>
> One issue with using ldap:// urls, if we do this I'm sure people will
> want to use all features of this url, and not understand that it is only
> meant to be an host + port. ldap://host:port/DN?attributes?scope?filter?
> extensions.
>
> Also, what about unix sockets, are these commonly used?
>
> How do you feel about a single option that holds either ssl/tls or
> start_tls?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/influxdata/telegraf/pull/2612#issuecomment-311186932>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAG57I7ZA9i6MC1RvIUVPmOTznn4m_0Wks5sICIpgaJpZM4Mxof9>
> .
>



-- 
andy

danielnelson commented:
So in Grafana, StartTLS doesn't do anything unless UseSSL is set, this seems a bit unexpected to me... the single option should resolve any confusion around this.

cobaugh commented:
Made ssl option a string, accepting "starttls" or "ldaps". Empty string disables encryption entirely. Any other value is an error. We also support ldapS now.

Also synced up with master and switched from HasIntField to HasInt64Field due to #2813

danielnelson commented:
Can you update the example output now that we are using integers?

cobaugh commented:
Example output and example config updated, as well as a couple of other minor issues fixed.

danielnelson commented:
This has been merged for 1.4, github just doesn't realize it.
